third_party/pyth/p2w-sdk/rust/src/lib.rs

```diff
@@ -80,28 +86,29 @@ pub enum PayloadId {
 #[serde(rename_all = "camelCase")]
 pub struct PriceAttestation {
     #[serde(serialize_with = "pubkey_to_hex")]
-    pub product_id:         Identifier,
+    pub product_id:            Identifier,
     #[serde(serialize_with = "pubkey_to_hex")]
-    pub price_id:           Identifier,
+    pub price_id:              Identifier,
     #[serde(serialize_with = "use_to_string")]
-    pub price:              i64,
+    pub price:                 i64,
     #[serde(serialize_with = "use_to_string")]
-    pub conf:               u64,
-    pub expo:               i32,
+    pub conf:                  u64,
+    pub expo:                  i32,
     #[serde(serialize_with = "use_to_string")]
-    pub ema_price:          i64,
+    pub ema_price:             i64,
     #[serde(serialize_with = "use_to_string")]
-    pub ema_conf:           u64,
-    pub status:             PriceStatus,
-    pub num_publishers:     u32,
-    pub max_num_publishers: u32,
-    pub attestation_time:   UnixTimestamp,
-    pub publish_time:       UnixTimestamp,
-    pub prev_publish_time:  UnixTimestamp,
+    pub ema_conf:              u64,
+    pub status:                PriceStatus,
+    pub num_publishers:        u32,
+    pub max_num_publishers:    u32,
+    pub attestation_time:      UnixTimestamp,
+    pub publish_time:          UnixTimestamp,
+    pub prev_publish_time:     UnixTimestamp,
     #[serde(serialize_with = "use_to_string")]
-    pub prev_price:         i64,
+    pub prev_price:            i64,
     #[serde(serialize_with = "use_to_string")]
-    pub prev_conf:          u64,
+    pub prev_conf:             u64,
+    pub prev_attestation_time: UnixTimestamp,
```

Comment:
I think we need to use a timestamp derived from the publish time not the attestation time for consistency. All the downstream SDKs use the publish time as the timestamp of the price update.

This will probably get confusing since there's already a field called `prev_publish_time`. However, I think this field should be something like `previously_attested_publish_time` which is set to either `publish_time` or `prev_publish_time` depending on whether the current status is trading or not (?). 

I think that ensures that there's at most one update per time interval, and that each interval ends with a usable price. (If the feed goes offline, I think you get a bunch of price updates where the current update timestamp is equal to the `previously_attested_publish_time`. When the feed comes back online, you get an update that spans the entire interval with the next available price.)

Please do check that logic and validate that it has the right properties. It's a little tricky.

Reply:
I like to suggest removing prev_* fields entirely and handling their logic directly here. We can do it in a way that it's not breaking (just pass 0 for those fields and replace one field with the field that we want)

Reply:
We should either pick what ambiguity we're more comfortable with, or simply use both values. The naming confusion you mention also bugs me a little.

Reply:
The name for the timestamp stored on chain is "last_attested_publish_time" to distinguish from existing `PriceAttestation` fields.